internal/reflect/slice.go

```diff
@@ -90,14 +90,71 @@ func reflectSlice(ctx context.Context, typ attr.Type, val tftypes.Value, target
 		}
 
 		return slice, diags
-	// Tuple reflection from slices is currently not supported as usage is limited
+	// Tuple reflection into slices is currently limited to use-cases where all tuple element types are the same.
 	case attr.TypeWithElementTypes:
-		diags.Append(diag.WithPath(path, DiagIntoIncompatibleType{
-			Val:        val,
-			TargetType: target.Type(),
-			Err:        fmt.Errorf("cannot reflect %s using type information provided by %T, reflection support is currently not implemented for tuples", val.Type(), t),
-		}))
-		return target, diags
+		// we need to know the type the slice is wrapping
+		elemType := target.Type().Elem()
+
+		// we want an empty version of the slice
+		slice := reflect.MakeSlice(target.Type(), 0, len(values))
+
+		if len(t.ElementTypes()) <= 0 {
+			// If the tuple values are empty as well, we can just pass back an empty slice of the type we received.
+			if len(values) == 0 {
+				return slice, diags
+			}
+
+			diags.Append(diag.WithPath(path, DiagIntoIncompatibleType{
+				Val:        val,
+				TargetType: target.Type(),
+				Err:        fmt.Errorf("cannot reflect %s using type information provided by %T, tuple type contained no element types but received values", val.Type(), t),
+			}))
+			return target, diags
+		}
+
+		// Ensure that all tuple element types are the same by comparing each element type to the first
+		multipleTypes := false
+		allElemTypes := t.ElementTypes()
+		elemAttrType := allElemTypes[0]
+		for _, elemType := range allElemTypes[1:] {
+			if !elemAttrType.Equal(elemType) {
+				multipleTypes = true
+				break
+			}
+		}
+
+		if multipleTypes {
+			diags.Append(diag.WithPath(path, DiagIntoIncompatibleType{
+				Val:        val,
+				TargetType: target.Type(),
+				Err:        fmt.Errorf("cannot reflect %s using type information provided by %T, reflection support for tuples is limited to multiple elements of the same element type. Expected all element types to be %T", val.Type(), t, elemAttrType),
+			}))
+			return target, diags
+		}
+
+		// go over each of the values passed in, create a Go value of the right
+		// type for them, and add it to our new slice
+		for pos, value := range values {
+			// create a new Go value of the type that can go in the slice
+			targetValue := reflect.Zero(elemType)
+
+			// update our path so we can have nice errors
+			// utilizing `AtListIndex` because tuple paths are also based on the underlying `PathStepElementKeyInt`
+			valPath := path.AtListIndex(pos)
```

Comment:
I have been using path.AtListIndex, since the underlying element key is the same for tuples, would it be confusing if I created a new path.AtTupleIndex equivalent for readability?

Reply:
I personally don't have a strong opinion either which way! Maybe we should create it so developers don't need to do the mental gymnastics of list indexing being equivalent to tuple indexing. 👍

Reply:
Cool, easy add then! Added `path.AtTupleIndex` in 8c5175a